compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,203 @@
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let predecessors = body.basic_blocks.predecessors();
+    let borrowed_locals = borrowed_locals(body);
+    // Should this be an `IndexVec` instead?
+    let mut new_targets = FxHashMap::default();
+    for (block, data) in body.basic_blocks.iter_enumerated() {
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        // To avoid possible writes to any parent projections of this local,
+        // we only allow this optimization on locals with no projections.
+        let Some(switched) = switched.as_local() else {
+            continue
+        };
+
+        // We also do not allow locals that are borrowed as they could be
+        // indirectly mutated before the switch.
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some((discr, index)) = data.statements.iter().enumerate().rev().find_map(
+            |(index, statement)| {
+                match statement.kind {
+                    StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                        if place.as_local() == Some(switched) =>
+                    {
+                        Some((discr, index))
+                    }
+                    _ => None,
+                }
+            }
+        ) else {
+            continue
+        };
+
+        let Some(discr) = discr.as_local() else {
+            continue
+        };
+
+        if borrowed_locals.contains(discr) {
+            continue;
+        }
+
+        // We need to make sure that `switched` is not overwritten before the switch,
+        // and also that `discr` is not overwritten before the `Rvalue::Discriminant`.
+        if is_local_mutated(body, block, discr, index, 0..index)
+            || is_local_mutated(body, block, switched, index, index..data.statements.len())
+        {
+            continue;
+        }
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
+                    StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place.as_local() == Some(discr) => {
```

Comment:
You need to check if `discr` is mutated between this assignment and the end of block.

Reply:
I have added a check for this.